arangod/Agency/AgencyComm.cpp

```diff
@@ -447,55 +435,72 @@ int AgencyCommResult::httpCode() const { return _statusCode; }
 bool AgencyCommResult::sent() const { return _sent; }
 
 int AgencyCommResult::errorCode() const {
-  try {
-    if (!_body.empty()) {
-      std::shared_ptr<VPackBuilder> bodyBuilder = VPackParser::fromJson(_body);
-      VPackSlice body = bodyBuilder->slice();
-      if (!body.isObject()) {
-        return 0;
-      }
-      // get "errorCode" attribute (0 if not exist)
-      return basics::VelocyPackHelper::getNumericValue<int>(body, "errorCode", 0);
-    }
-  } catch (VPackException const&) {
-  }
-  return 0;
+  return asResult().errorNumber();
 }
 
 std::string AgencyCommResult::errorMessage() const {
-  if (!_message.empty()) {
-    // return stored message first if set
-    return _message;
-  }
-
-  if (!_connected) {
-    return std::string("unable to connect to agency");
-  }
+  return asResult().errorMessage();
+}
 
+std::optional<std::pair<int, std::string_view>> AgencyCommResult::parseBodyError() const {
+  auto result = std::optional<std::pair<int, std::string_view>>{};
   try {
-    std::shared_ptr<VPackBuilder> bodyBuilder = VPackParser::fromJson(_body);
-
-    VPackSlice body = bodyBuilder->slice();
-    if (!body.isObject()) {
-      return "";
+    if (_vpack != nullptr) {
+      auto const body = _vpack->slice();
+      if (body.isObject()) {
+        // get "errorCode" attribute
+        auto const errorCode = body.get(StaticStrings::ErrorCode).getNumber<int>();
+        // Save error code if possible, set default error message first
+        result = std::make_pair(errorCode, std::string_view(TRI_errno_string(errorCode)));
+        // Now try to extract the message, too; but it's fine if that fails, we
+        // already have the default one.
+        if (auto const errMsg = body.get(StaticStrings::ErrorMessage); errMsg.isString()) {
+          result->second = errMsg.stringView();
+        } else if (auto const errMsg = body.get("message"); errMsg.isString()) {
+          result->second = errMsg.stringView();
+        }
+      }
     }
-    // get "message" attribute ("" if not exist)
-    return arangodb::basics::VelocyPackHelper::getStringValue(body, "message",
-                                                              "");
-  } catch (VPackException const& e) {
-    std::string message("VPackException parsing body (" + _body + "): " + e.what());
-    return std::string(message);
+  } catch (VPackException const&) {
   }
+
+  return result;
 }
 
 std::string AgencyCommResult::errorDetails() const {
-  std::string const errorMessage = this->errorMessage();
+  auto const errorMessage = this->errorMessage();
 
   if (errorMessage.empty()) {
     return _message;
   }
 
-  return _message + " (" + errorMessage + ")";
+  return _message + " (" + std::string{errorMessage} + ")";
```

Comment:
AFAIK `errorMessage` is already a string here, so no need to put it into another string temporary.
```suggestion
  return _message + " (" + errorMessage + ")";
```